src/serac/physics/integrators/traction_integrator.cpp

```diff
@@ -0,0 +1,226 @@
+                                            const mfem::Vector&               input_state_vector,
+                                            mfem::Vector&                     output_residual_vector)
+{
+  // Get the dimension and number of degrees of freedom from the current element
+  int dim = element_1.GetDim();
+  int dof = element_1.GetDof();
+
+  // Ensure the working data structures are sized appropriately
+  shape_.SetSize(dof);
+  output_residual_vector.SetSize(dim * dof);
+  dN_dxi_.SetSize(dof, dim);
+  dN_dX_.SetSize(dof, dim);
+  dxi_dX_.SetSize(dim);
+  F_.SetSize(dim);
+  Finv_.SetSize(dim);
+  traction_vector_.SetSize(dim);
+  reference_normal_.SetSize(dim);
+  current_normal_.SetSize(dim);
+
+  // Reshape the input state as a matrix (dof x dim)
+  input_state_matrix_.UseExternalData(input_state_vector.GetData(), dof, dim);
+
+  // Calculate an appropriate integration rule for the element order
+  int                          intorder = 2 * element_1.GetOrder() + 3;
+  const mfem::IntegrationRule& ir = mfem::IntRules.Get(parent_to_reference_face_transformation.FaceGeom, intorder);
+
+  // Initialize the residual output
+  output_residual_vector = 0.0;
+
+  for (int i = 0; i < ir.GetNPoints(); i++) {
+    // Set the current integration point
+    const mfem::IntegrationPoint& ip = ir.IntPoint(i);
+    mfem::IntegrationPoint        eip;
+    parent_to_reference_face_transformation.Loc1.Transform(ip, eip);
+    parent_to_reference_face_transformation.Face->SetIntPoint(&ip);
+
+    // Compute the traction at the integration point
+    traction_.Eval(traction_vector_, *parent_to_reference_face_transformation.Face, ip);
+
+    // If computing on the deformed configuration, transform the normal vector appropriately
+    if (!compute_on_reference_) {
+      // Calculate the deformation gradient and its inverse
+      parent_to_reference_face_transformation.Elem1->SetIntPoint(&eip);
+      CalcInverse(parent_to_reference_face_transformation.Elem1->Jacobian(), dxi_dX_);
+      element_1.CalcDShape(eip, dN_dxi_);
+      Mult(dN_dxi_, dxi_dX_, dN_dX_);
+      MultAtB(input_state_matrix_, dN_dX_, F_);
+
+      for (int d = 0; d < dim; d++) {
+        F_(d, d) += 1.0;
+      }
+
+      CalcInverse(F_, Finv_);
+
+      // Calculate the normal vector in the reference configuration
+      CalcOrtho(parent_to_reference_face_transformation.Face->Jacobian(), reference_normal_);
+
+      // Normalize vector
+      double norm = reference_normal_.Norml2();
+      reference_normal_ /= norm;
+
+      // Calculate F inverse time the normal vector
+      Finv_.MultTranspose(reference_normal_, current_normal_);
+    }
+
+    // Calculate the shape functions
+    element_1.CalcShape(eip, shape_);
+    for (int j = 0; j < dof; j++) {
+      for (int k = 0; k < dim; k++) {
+        double residual_contribution =
+            -1.0 * traction_vector_(k) * shape_(j) * ip.weight * parent_to_reference_face_transformation.Face->Weight();
```

Comment:
Do you want the -1.0 here?

Reply:
Yes - this is due to a sign factor the MFEM assembly routine. This is the contribution to the residual (R(u) = K(u) + t(u) = 0) vs the RHS (K(u) = t(u)).